### PR TITLE
Support callback as data source

### PR DESCRIPTION
**Pull request by xinyifly**
I initially need to add an `Authorization` header contains a bearer token. Since my project is already using a configured `axios`, supporting callback as data source would solve my problem and provide most flexibility meanwhile.

### CODE REVIEW

**Comment by ankane** [comment]
Hey @xinyifly, thanks for the PR 👍 Overall, looks great. Can you rewrite the first part without promises for max compatibility?

**Comment by xinyifly** [comment]
Hi @ankane , if the motivation is removing promise from the library at all, I can rewrite it to receive pure callback, and the async life seems will be a bit harder...

But if we just want to support pure callback besides promise, this code already did it.
```JavaScript
Promise.resolve('Pure Text').then(function (data) { console.log(data) })  // Pure Text
```

Let me know how to amend it!

**Comment by ankane** [comment]
@xinyifly the library shouldn't use promises internally. I'm thinking something like this:

```js
let cb = function(data) {
  chart.rawData = data;
  errorCatcher(chart);
};

try {
  dataSource(cb);
} catch (err) {
  chartError(chart.element, err);
}
```

And users could pass:

```js
function myDataSource(cb) {
  // do something
  cb(data);
}
```

**Comment by xinyifly** [comment]
@ankane Yes, I'm thinking about calling `dataSource(cb)` also, doubting if you can catch the error if it executed asynchronously, tho. Should it be `dataSource(cb, failCb)` or not?

**Comment by ankane** [comment]
@xinyifly Good point about also providing an error callback. That pattern makes sense to me. An alternative could be a single callback (error-first callback pattern) with:

```js
cb(null, data)
cb("some error")
```

But I think I like your proposal better.

File: src/index.js [hunk 1]
```diff
@@ -386,14 +393,16 @@ class Chart {
...
+      fetchDataSource(this, this.dataSource);
     }
   }
 
   startRefresh() {
     let refresh = this.options.refresh;
 
-    if (refresh && typeof this.dataSource !== "string") {
-      throw new Error("Data source must be a URL for refresh");
+    if (refresh && typeof this.dataSource !== "string" && typeof this.dataSource !== "function") {
```
[the remark this answers is not in the document]
Comment (ankane) [reply src/index.js]: nvm, you have it right

**Comment by ankane** [comment]
Looks great, big thanks @xinyifly 🎉

**Pull request merged**